frontend/src/stimulus/controllers/pattern-input.controller.ts

```diff
@@ -395,28 +395,36 @@ export default class PatternInputController extends Controller {
   private createToken(value:string):HTMLElement {
     const templateTarget = this.tokenTemplateTarget.content?.cloneNode(true) as HTMLElement;
     const contentElement = templateTarget.firstElementChild as HTMLElement;
-    contentElement.innerText = value;
+    contentElement.dataset.prop = value;
+    contentElement.innerText = this.validTokenMap[value];
     return contentElement;
   }
 
   private toHtml(blueprint:string):string {
-    return blueprint.replace(/{{([0-9A-Za-z_]+)}}/g, (_, token:string) => this.createToken(token).outerHTML);
+    return blueprint
+      .replace(/</g, '&lt;')
+      .replace(/>/g, '&gt;')
+      .replace(/{{([0-9A-Za-z_]+)}}/g, (_, token:string) => this.createToken(token).outerHTML);
   }
 
   private toBlueprint():string {
     let result = '';
-    this.contentTarget.childNodes.forEach((node:Element) => {
+    this.contentTarget.childNodes.forEach((node:ChildNode) => {
       if (node.nodeType === Node.TEXT_NODE) {
         // Plain text node
         result += node.textContent;
-      } else if (node.nodeType === Node.ELEMENT_NODE && (node as HTMLElement).dataset.role === 'token') {
+      } else if (this.isToken(node)) {
         // Token element
-        result += `{{${node.textContent?.trim()}}}`;
+        result += `{{${(node as HTMLElement).dataset.prop}}}`;
       }
     });
     return result.trim();
   }
 
+  private isToken(node:ChildNode):boolean {
+    return node.nodeType === Node.ELEMENT_NODE && (node as HTMLElement).dataset.role === 'token';
+  }
```

Comment:
```suggestion
  private isToken(node:ChildNode):node is TokenElement {
    return this.isElement(node) && node.dataset.role === 'token';
  }

  private isText(node:Node):node is Text {
    return node.nodeType === Node.TEXT_NODE; 
  }

  private isElement(node:Node):node is HTMLElement {
    return node.nodeType === Node.ELEMENT_NODE;
  }
```

Reply:
(`isText` and `isElement` might be better defined as utility functions somewhere - and for that matter, `isWhitespace` too)

Reply:
IMHO the functions are too small and too trivial yet, to be extracted into utility functions.